internal/proxy/oauthproxy.go

```diff
@@ -655,6 +663,12 @@ func (p *OAuthProxy) Proxy(rw http.ResponseWriter, req *http.Request) {
 			// the user has a stale sesssion.
 			p.OAuthStart(rw, req, tags)
 			return
+		case ErrUnauthorizedUpstreamRequested:
+			// The users session has been authorised for use with a different upstream than the one
+			// that is being requested, so we trigger the start of the oauth flow.
+			// This exists primarily to implement some form of grace period while this additional session
+			// check is being introduced.
+			p.OAuthStart(rw, req, tags)
```

Comment:
Question: will this invalidate the current validated session for 'foo' upstream?

Reply:
It doesn't invalidate the session on the original upstream - and because it restarts the oauth flow for the _new_ request/upstream the UX should be pretty seamless when this is triggered.